# Practice Session 04: Basket analysis

Association rule mining techniques are useful to analyze datasets consisting of transactions, in which each transaction is a collection of items.

We will use a well-known dataset named [Instacart](https://www.kaggle.com/c/instacart-market-basket-analysis) containing more than 3 million orders of products through a grocery shopping app. You can find it in the `instacart/` directory of the practicum data files.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

Author: <font color="blue">Bernat Quintilla Castellón</font>

E-mail: <font color="blue">bernat.quintilla01@estudiant.upf.edu</font>

Date: <font color="blue">The current date here</font>

In [1]:
import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd  
import csv
import gzip
from apyori import apriori

If the apyori library is not already installed in your laptop, you can install it with: `pip install apyori`

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

## 0. The Apriori Algorithm in a nutshell

There are three major components of Apriori algorithm, which we describe below using as an example the case where transactions are purchase histories.

**Support**: the number of transactions containing a particular item divided by total number of transactions:

   *Support(A) = (Transactions containing (A))/(Total Transactions)*

**Confidence**: normally indicates the likelihood that an item B is also bought if item A is bought. It can be calculated by finding the number of transactions where A and B are bought together, divided by total number of transactions where A is bought:

   *Confidence(A→B) = (Transactions containing both (A and B))/(Transactions containing A)*

**Lift**: the increase in the ratio of sale of B when A is sold. Lift(A –> B) can be calculated by dividing Confidence(A -> B) by Support(B):

   *Lift(A→B) = (Confidence (A→B))/(Support (B))*
   
A Lift of 1 means there is no association between products A and B. Lift greater than 1.0 means products A and B are more likely to be bought together. Lift less than 1.0 indicates two products are unlikely to be bought together.

The Apriori algorithm first finds itemsets having the desired level of support, and then within those itemsets tries to derive rules having the desired confidence and lift.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

# 1. Playing with apyori

The [apyori library](https://pypi.org/project/apyori/) is an implementation of the Apriori algorithm. Its typical usage is to receive a list of transactions and then print the association rules it found.

To use this library, we pass a list in which each element represents a transaction, for instance:

```python
transactions = [
    ['beer', 'chips', 'nuts', 'olives'],
    ['beer', 'chips', 'olives'],
    ['chips', 'nuts' ],
    ['chips', 'olives'],
    ['beer', 'nuts' ],
    ['chips'],
    ['nuts', 'olives'],
    ['beer', 'nuts'],
    ['beer', 'chips', 'olives'], 
    ['beer', 'nuts', 'olives'], 

]
results = list(apriori(transactions, min_support=0.3, min_confidence=0.75, min_lift=1.0))

```

The function below, which you can leave as-is, prints the output of the apyori library in a readable format. Use it to print the results of your association rules mining:

```python
print_apyori_output(results)
```

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

In [2]:
# LEAVE AS-IS

def print_apyori_output (association_results, info=False, info_key=False):
    for relation_record in association_results:
        itemset = list(relation_record.items)
        
        # Consider only itemsets of two elements
        if len(itemset) > 1: 
        
            print("Rules involving itemset %s" % itemset)
            support = relation_record.support

            for rules in relation_record.ordered_statistics:
                antecedent = list(rules.items_base)
                consequent = list(rules.items_add)
                
                if info_key:
                    antecedent = [info.loc[x][info_key] for x in antecedent]
                    consequent = [info.loc[x][info_key] for x in consequent]
                
                confidence = rules.confidence
                lift = rules.lift

                print("%s => %s (support=%.4f, confidence=%.2f, lift=%.2f)" %
                      (antecedent, consequent, support, confidence, lift))
            print()

<font size="+1" color="red">Replace this cell with your own example of transactions (at least 20 transactions) and execution of the apriori algorithm, in which you should obtain at least ONE and at most THREE rules.</font>

In [3]:
#21 different transactions
transactions = [
    ['milk', 'bread', 'eggs', 'cheese'],
    ['bread', 'eggs', 'cheese'],
    ['milk', 'eggs', 'cheese'],
    ['milk', 'bread', 'cheese'],
    ['milk', 'bread', 'eggs'],
    ['bread', 'cheese'],
    ['milk', 'eggs'],
    ['milk', 'bread'],
    ['bread', 'eggs'],
    ['milk', 'cheese'],
    ['eggs'],
    ['bread', 'butter'],
    ['milk', 'butter'],
    ['butter'],
    ['milk', 'yogurt'],
    ['yogurt', 'eggs'],
    ['milk', 'yogurt', 'eggs'],
    ['yogurt', 'cheese'],
    ['milk', 'yogurt', 'cheese'],
    ['yogurt', 'bread'],
    ['milk', 'yogurt', 'bread']
]

results = list(apriori(transactions, min_support=0.1, min_confidence=0.5, min_lift=1.0))

print_apyori_output(results)

Rules involving itemset ['bread', 'cheese']
['cheese'] => ['bread'] (support=0.1905, confidence=0.50, lift=1.05)

Rules involving itemset ['cheese', 'milk']
['cheese'] => ['milk'] (support=0.2381, confidence=0.62, lift=1.09)

Rules involving itemset ['yogurt', 'milk']
['yogurt'] => ['milk'] (support=0.1905, confidence=0.57, lift=1.00)



<font size="+1" color="red">Replace this cell with (1) a printout of the rules you have obtained, and (2) for each of those rules, indicate clearly how the support, confidence, and lift is calculated. Do not merely repeat the formula: indicate how each number is computed based on the transactions you provided, as if you were trying to verify that the numbers are correct.</font>

**Rule of itemset ['bread', 'cheese']: ['cheese'] => ['bread']**

```Support```:

Support = Number of transactions with ['cheese', 'bread'] / Total number of transactions

Support = 4 / 20 = 0.1905

```Confidence```:

Confidence = Number of transactions with ['cheese', 'bread'] / Number of transactions with ['cheese']

Confidence = 4 / 8 = 0.50

```Lift```:

Lift = Support / (Support of 'cheese' * Support of 'bread')

Lift = 0.1905 / (0.4286 * 0.4762) = 1.05


**Rule of itemset ['cheese', 'milk']: ['cheese'] => ['milk']**

```Support```:

Support = Number of transactions with ['cheese', 'milk'] / Total number of transactions

Support = 5 / 20 = 0.2381

```Confidence```:

Confidence = Number of transactions with ['cheese', 'milk'] / Number of transactions with ['cheese']

Confidence = 5 / 8 = 0.62

```Lift```:

Lift = Support / (Support of 'cheese' * Support of 'milk')

Lift = 0.2381 / (0.4286 * 0.5238) = 1.09

**Rule of itemset ['yogurt', 'milk']: ['yogurt'] => ['milk']**

```Support```:

Support = Number of transactions with ['yogurt', 'milk'] / Total number of transactions

Support = 4 / 20 = 0.1905

```Confidence```:

Confidence = Number of transactions with ['yogurt', 'milk'] / Number of transactions with ['yogurt']

Confidence = 4 / 7 = 0.57

```Lift```:

Lift = Support / (Support of 'yogurt' * Support of 'milk')

Lift = 0.1905 / (0.3333 * 0.5714) = 1.00

# 2. Load and prepare the shopping baskets

The following code, which you should leave as-is, loads the information about products into a dataframe indexed by product id.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

In [4]:
# LEAVE AS-IS

# File names
INPUT_PRODUCTS = "instacart-products.csv"
INPUT_TRANSACTIONS = "instacart-transactions.csv.gz"

# Read into a dataframe
products = pd.read_csv(INPUT_PRODUCTS, delimiter=",")

# Set product_id as index, and drop column aisle_id
products = products.set_index('product_id').drop(columns=['aisle_id'])

products.head(100)

product_name  department_id
product_id                                                                  
1                                  Chocolate Sandwich Cookies             19
2                                            All-Seasons Salt             13
3                        Robust Golden Unsweetened Oolong Tea              7
4           Smart Ones Classic Favorites Mini Rigatoni Wit...              1
5                                   Green Chile Anytime Sauce             13
...                                                       ...            ...
96                    Sprinklez Confetti Fun Organic Toppings             13
97                                Organic Chamomile Lemon Tea              7
98                                  2% Yellow American Cheese             16
99                                Local Living Butter Lettuce              4
100                   Peanut Butter & Strawberry Jam Sandwich              1

[100 rows x 2 columns]

## 2.1. Select by department

As this file is large and complex, we will focus on one or two departments and try to get some conclusions about the products in those departments. The following cell, which you should leave as-is, list some department names.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

In [5]:
# LEAVE AS-IS

DEPT_BAKERY = 3
DEPT_VEGGIES = 4
DEPT_ALCOHOL = 5
DEPT_WORLD = 6
DEPT_DRINKS = 7
DEPT_PETS = 8
DEPT_PHARMACY = 11
DEPT_CLEANING = 17
DEPT_BABIES = 18

Write code that can select a list of products from a set of departments. Do this with a function named `select_from_departments` that takes as input:

* A dataframe containing product information, which will be the `products` dataframe we just loaded.
* A list of product ids
* A list of department ids

It should return a list containing only the product ids that belong to one of the listed departments. This may return an empty list if no product belongs to any of the specified departments.

Given that the products dataframe is indexed by *product_id*, if you want to obtain the *department_id* of product *product_id*, use:

```python
products.loc[product_id].department_id
```

Note that *product_id* must be an integer.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

<font size="+1" color="red">Replace this cell with your code for *select_from_departments*.</font>

In [6]:
ids_product = list(range(50))

def select_from_departments(products, product_ids, department_ids):
    selected_products = []

    for product_id in product_ids:
        if product_id in products.index:
            department_id = products.loc[product_id].department_id
            if department_id in department_ids:
                selected_products.append(product_id)

    return selected_products

selected_products = select_from_departments(products, ids_product, [DEPT_WORLD, DEPT_PETS])
print(selected_products)

[21, 26, 47]


Test your function by passing it a list of products and ensuring it selects only the products in the 1-2 departments you have selected. To obtain test cases you can open the products file with a spreadsheet program.

Each test case should print:

* The product name and department id of each item in the input list
* The product name and department id of each item in the output list

For instance, suppose a test case is `[22, 26, 45, 54, 57, 71, 111, 112]` and we select products from DEPT_BAKERY and DEPT_CLEANING, a test case run should print something similar to this:

```
Test case:
[22, 26, 45, 54, 57, 71, 111, 112]

Input products:
22 Fresh Breath Oral Rinse Mild Mint (dept 11)
26 Fancy Feast Trout Feast Flaked Wet Cat Food (dept 8)
45 European Cucumber (dept 4)
54 24/7 Performance Cat Litter (dept 8)
57 Flat Toothpicks (dept 17)
71 Ultra 7 Inch Polypropylene Traditional Plates (dept 17)
111 Fabric Softener, Geranium Scent (dept 17)
112 Hot Tomatillo Salsa (dept 13)

Selected products:
57 Flat Toothpicks (dept 17)
71 Ultra 7 Inch Polypropylene Traditional Plates (dept 17)
111 Fabric Softener, Geranium Scent (dept 17)
```

Do not replicate code that can be easily factored in a function in your answer.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

<font size="+1" color="red">Replace this cell with code to test your function with three different test cases. Each test case is a list of items and a list of 1, 2, or 3 departments.</font>

In [7]:
id_products_test1 = [22, 26, 45, 54, 57, 71, 111, 112]
id_products_test2 = [12, 32, 45, 63, 71, 77, 85, 90, 110]
id_products_test3 = [8, 14, 23, 26, 33, 38, 46, 50, 55, 63, 73, 87, 92, 104]

list_dept_test1 = [DEPT_BAKERY, DEPT_CLEANING]
list_dept_test2 = [DEPT_VEGGIES, DEPT_ALCOHOL, DEPT_DRINKS]
list_dept_test3 = [DEPT_PHARMACY, DEPT_BABIES, DEPT_WORLD]

def print_test_selected_products(products, id_products, list_dept):
    product_df = pd.DataFrame(products, index = id_products)
    selected_products = select_from_departments(product_df, id_products, list_dept)
    print("Test case with product ids: ",id_products," and departments: ",list_dept,"\n",selected_products,"\n")

print_test_selected_products(products, id_products_test1, list_dept_test1)
print_test_selected_products(products, id_products_test2, list_dept_test2)
print_test_selected_products(products, id_products_test3, list_dept_test3)

Test case with product ids:  [22, 26, 45, 54, 57, 71, 111, 112]  and departments:  [3, 17] 
 [57, 71, 111] 

Test case with product ids:  [12, 32, 45, 63, 71, 77, 85, 90, 110]  and departments:  [4, 5, 7] 
 [45] 

Test case with product ids:  [8, 14, 23, 26, 33, 38, 46, 50, 55, 63, 73, 87, 92, 104]  and departments:  [11, 18, 6] 
 [55, 63] 



## 2.2. Read and filter transactions

The transactions file is a compressed file containing one row per transaction. Each transaction is a comma-separated list of *product_id*. The following skeleton iterates through this file:

```python
# Open a compressed file
with gzip.open(INPUT_TRANSACTIONS, "rt") as inputfile:
    
    # Create a CSV reader
    reader = csv.reader(inputfile, delimiter=",")
    
    # Iterate through the CSV file
    for row in reader:
        
        # Convert to integers
        items = [int(x) for x in row]
```

Read the transactions, filtering the items so they belong to the DEPT_CLEANING department. Stop reading (`break`) after you have stored 5000 transactions into an array named `transactions`. Every 1000 transactions read, print the number of transactions read and the number of transactions stored.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

<font size="+1" color="red">Replace this cell with your code to read transactions, keeping only items in DEPT_CLEANING. Remember to stop after keeping 5000 transactions.</font>

In [15]:
transactions_read = 0
transactions_stored = 0
transactions=[]
# Open a compressed file
with gzip.open(INPUT_TRANSACTIONS, "rt") as inputfile:

    # Create a CSV reader
    reader = csv.reader(inputfile, delimiter=",")

    # Iterate through the CSV file
    for row in reader:

        # Convert to integers
        items = [int(x) for x in row]
        selected_items = select_from_departments(products, items, [DEPT_CLEANING]) #Select products_id in items that are in DEPT_CLEANING
        if selected_items: #if it is not empty
            transactions.append(items)
            transactions_stored += 1

        transactions_read += 1

        if transactions_stored >= 5000:#Check if we've stored 5000 transactions
            break

        if transactions_read % 1000 == 0:#Print progress every 1000 transactions
            print(f"Transactions Read: {transactions_read}, Transactions Stored: {transactions_stored}")

print("Finished reading and storing transactions.")

Transactions Read: 1000, Transactions Stored: 158
Transactions Read: 2000, Transactions Stored: 311
Transactions Read: 3000, Transactions Stored: 460
Transactions Read: 4000, Transactions Stored: 598
Transactions Read: 5000, Transactions Stored: 745
Transactions Read: 6000, Transactions Stored: 902
Transactions Read: 7000, Transactions Stored: 1067
Transactions Read: 8000, Transactions Stored: 1206
Transactions Read: 9000, Transactions Stored: 1373
Transactions Read: 10000, Transactions Stored: 1515
Transactions Read: 11000, Transactions Stored: 1670
Transactions Read: 12000, Transactions Stored: 1807
Transactions Read: 13000, Transactions Stored: 1951
Transactions Read: 14000, Transactions Stored: 2102
Transactions Read: 15000, Transactions Stored: 2245
Transactions Read: 16000, Transactions Stored: 2384
Transactions Read: 17000, Transactions Stored: 2543
Transactions Read: 18000, Transactions Stored: 2692
Transactions Read: 19000, Transactions Stored: 2840
Transactions Read: 20000, T

## 2.3. Extract association rules and comment on them (DEPT_CLEANING)

You are now ready to run the association rules mining algorithm over the selected transactions:

```python
results = list(apriori(transactions, min_support=..., min_confidence=..., min_lift=...))
print_apyori_output(results, products, 'product_name')
```

*Tip:* if you set `min_support` to a very small value, your notebook will probably hang.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

<font size="+1" color="red">Replace this cell with your code to extract association rules from the read transactions (DEPT_CLEANING).</font>

In [30]:
results = list(apriori(transactions, min_support=0.003, min_confidence=0.5, min_lift=1.0))
print_apyori_output(results, products, 'product_name')

Rules involving itemset [19604, 16797]
['Medium Scarlet Raspberries'] => ['Strawberries'] (support=0.0030, confidence=0.56, lift=12.35)

Rules involving itemset [13176, 5876, 27966]
['Organic Lemon', 'Organic Raspberries'] => ['Bag of Organic Bananas'] (support=0.0030, confidence=0.71, lift=6.24)

Rules involving itemset [13176, 47209, 8021]
['Organic Hass Avocado', '100% Recycled Paper Towels'] => ['Bag of Organic Bananas'] (support=0.0042, confidence=0.52, lift=4.59)

Rules involving itemset [13176, 21137, 27966]
['Organic Strawberries', 'Organic Raspberries'] => ['Bag of Organic Bananas'] (support=0.0058, confidence=0.60, lift=5.28)

Rules involving itemset [13176, 21137, 39275]
['Organic Strawberries', 'Organic Blueberries'] => ['Bag of Organic Bananas'] (support=0.0030, confidence=0.50, lift=4.37)

Rules involving itemset [13176, 39275, 27966]
['Organic Blueberries', 'Organic Raspberries'] => ['Bag of Organic Bananas'] (support=0.0030, confidence=0.63, lift=5.46)

Rules involving 

<font size="+1" color="red">Replace this cell with a brief commentary on what you would recommend to the shopping app considering the extracted association rules.</font>

I would recommend to the shopping app to offer bundle deals or discounts for items that appear together in these association rules. For instance, the rule ['Organic Lemon', 'Organic Raspberries'] => ['Bag of Organic Bananas'] suggests that customers who buy Organic Lemon and Organic Raspberries often also purchase Bag of Organic Bananas. The app can offer a discount when these three items are bought together for example.

Another recomendation would be that because of the fact that certain items are frequently bought together, the app should promote the related categories as well. For example, if 'Organic Strawberries,' 'Organic Raspberries,' and 'Bag of Organic Bananas' are frequently bought together, the app should suggest to the customer the "Organic Fruits" category.

## 2.4. Extract association rules and comment on them (other departments)

<font size="+1" color="red">Replace this cell with code to select a different set of departments (at least two, not DEPT_CLEANING) and extract transactions again. Avoid replicating code when possible.</font>

<font size="+1" color="red">Replace this cell with your commentary on the obtained rules.</font>

# DELIVER (individually)

Remember to read the section on "delivering your code" in the [course evaluation guidelines](https://github.com/chatox/data-mining-course/blob/master/upf/upf-evaluation.md).

Deliver a zip file containing:

* This notebook

## Extra points available

For more learning and extra points, write code to filter the obtained association rules so that you print only the ones involving products in different departments. To be precise, this means rules in which there is at least a product in the *consequence* that belongs to a department that none of the products in the *antecedent* belongs to. Experiment with different combinations of departments, and try to discover interesting groups of products in different departments that are related to each other.

**Note:** if you go for the extra points, add ``<font size="+2" color="blue">Additional results: experiments on cross-department association rules</font>`` at the top of your notebook.

<font size="-1" color="gray">(Remove this cell when delivering.)</font>

<font size="+2" color="#003300">I hereby declare that, except for the code provided by the course instructors, all of my code, report, and figures were produced by myself.</font>